api/src/main/java/org/apache/iceberg/metrics/Timer.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.metrics;
+
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * Generalized Timer interface for creating telemetry related instances for measuring duration of operations.
+ */
+public interface Timer {
+
+  /**
+   * The number of times {@link Timer#record(Duration)} was called.
+   *
+   * @return The number of times {@link Timer#record(Duration)} was called.
+   */
+  long count();
+
+  /**
+   * The total duration that was recorded.
+   *
+   * @return The total duration that was recorded.
+   */
+  Duration totalDuration();
+
+  /**
+   * Starts the timer
+   */
+  void startTimer();
+
+  /**
+   * Stops the timer and records the total duration up until {@link Timer#startTimer()} was called.
+   */
+  void stopTimer();
+
+  /**
+   * Records a custom amount in the given time unit.
+   *
+   * @param amount The amount to record
+   * @param unit   The time unit of the amount
+   */
+  void record(long amount, TimeUnit unit);
+
+  /**
+   * The duration to record
+   *
+   * @param duration The duration to record
+   */
+  default void record(Duration duration) {
+    record(duration.toNanos(), TimeUnit.NANOSECONDS);
+  }
+
+  /**
+   * Executes and measures the given {@link Runnable} instance.
+   *
+   * @param runnable The {@link Runnable} to execute and measure.
+   */
+  void record(Runnable runnable);
+
+  /**
+   * Executes and measures the given {@link Callable} and returns its result.
+   *
+   * @param callable The {@link Callable} to execute and measure.
+   * @param <T>      The type of the {@link Callable}
+   * @return The result of the underlying {@link Callable}.
+   * @throws Exception In case the {@link Callable} fails.
+   */
+  <T> T recordCallable(Callable<T> callable) throws Exception;
```

Comment:
An alternative that I tend to like is to have a custom `Task` interface rather than using `Callable` that allows you to specify the exception type:

```java
  public interface TimedTask<R, E extends Exception> {
    R call() throws E;
  }
```

That way, you can customize the exception type and don't need an outer `try` block if it only throws `RuntimeException`.

Reply:
I like that idea. Would you like me to add this as part of this PR or rather later?

Reply:
I think it would also make sense to have something like a `TimedRunnable` / `TimedSupplier`

Reply:
I'd probably replace what's here with two variations, one that returns a value and one that has no return. Then you've covered all of the cases and can use method references to call with a Runnable, Supplier, or Callable.